File: Python/vovapin.py
import asyncio
import json
from datetime import datetime
from aiohttp import ClientSession, BasicAuth

# Конфигурация
PINNACLE_API_URL = "https://api.pinnacle.com/"
PINNACLE_USERNAME = "AG1677099"
PINNACLE_PASSWORD = "5421123A"
SPORT_ID = 29  # ID футбола в Pinnacle API
LIVE_MODE = 1  # Только лайв
ODDS_FORMAT = "Decimal"
REQUEST_TIMEOUT = 5
PROXY = "http://AllivanService:PinnacleProxy@154.7.188.227:5242"

# Глобальное хранилище данных о матчах
matches_data = {}

# Класс для работы с Pinnacle API
class PinnacleAPI:
    def __init__(self, username, password, proxy=None):
        self.username = username
        self.password = password
        self.proxy = proxy

    def _url(self, endpoint, params=None):
        url = f"{PINNACLE_API_URL}{endpoint}?"
        if params:
            url += "&".join(f"{k}={v}" for k, v in params.items() if v is not None)
        return url

    async def _query(self, url):
        try:
            async with ClientSession() as session:
                async with session.get(
                    url,
                    timeout=REQUEST_TIMEOUT,
                    proxy=self._get_proxy(),
                    auth=self._get_auth(),
                    headers=self._get_headers()
                ) as response:
                    print(f"[LOG] Запрос к URL: {url}")
                    print(f"[LOG] Статус ответа: {response.status}")
                    result = await response.text()
                    return json.loads(result)
        except Exception as e:
            print(f"[ERROR] Ошибка запроса: {e}")
            return None

    def _get_auth(self):
        return BasicAuth(self.username, self.password)

    def _get_headers(self):
        return {"Accept": "application/json", "Content-Type": "application/json"}

    def _get_proxy(self):
        return self.proxy


# Функция для обновления данных о матчах
async def fetch_matches(api: PinnacleAPI):
    global matches_data
    url = api._url("v1/fixtures", {
        "sportId": SPORT_ID,
        "isLive": LIVE_MODE
    })
    data = await api._query(url)

    if not data or "league" not in data:
        print("[ERROR] Не удалось загрузить данные о матчах.")
        return

    matches_data.clear()
    for league in data["league"]:
        league_name = league.get("name", "Неизвестная лига")
        for event in league.get("events", []):
            if event.get("liveStatus", 0) == 1:
                event_id = event["id"]
                matches_data[event_id] = {
                    "home": event.get("home", "Неизвестная команда"),
                    "away": event.get("away", "Неизвестная команда"),
                    "league": league_name,
                    "start_time": event.get("starts"),
                }
    #print("[DEBUG] matches_data после обновления:", json.dumps(matches_data, indent=2, ensure_ascii=False))


# Функция для обработки данных события
def process_match_data(event_data):
    try:
        event_id = event_data["id"]
        if event_id in matches_data:
            home_team = matches_data[event_id]["home"]
            away_team = matches_data[event_id]["away"]
            league_name = matches_data[event_id]["league"]
            start_time = matches_data[event_id]["start_time"]
        else:
            print(f"[WARNING] Событие {event_id} отсутствует в matches_data.")
            home_team = "Неизвестная команда"
            away_team = "Неизвестная команда"
            league_name = "Неизвестная лига"
            start_time = None

        if start_time:
            start_time = datetime.fromisoformat(start_time.replace("Z", "+00:00"))

        outcomes = []
        for period in event_data.get("periods", []):
            if period.get("status") != 1:
                continue

            moneyline = period.get("moneyline", {})
            if moneyline:
                outcomes.append({"market": "Победа", "team": "Home", "odds": moneyline.get("home")})
                outcomes.append({"market": "Победа", "team": "Draw", "odds": moneyline.get("draw")})
                outcomes.append({"market": "Победа", "team": "Away", "odds": moneyline.get("away")})

            totals = period.get("totals", [])
            for total in totals:
                outcomes.append({
                    "market": "Общий тотал",
                    "line": total.get("points"),
                    "over_odds": total.get("over"),
                    "under_odds": total.get("under"),
                })

            spreads = period.get("spreads", [])
            for spread in spreads:
                outcomes.append({
                    "market": "Гандикап",
                    "line": spread.get("hdp"),
                    "home_odds": spread.get("home"),
                    "away_odds": spread.get("away"),
                })

            team_totals = period.get("teamTotal", {})
            for team, total in team_totals.items():
                if not total:
                    continue
                team_name = "Home" if team == "home" else "Away"
                outcomes.append({
                    "market": "Индивидуальный тотал",
                    "team": team_name,
                    "line": total.get("points"),
                    "over_odds": total.get("over"),
                    "under_odds": total.get("under"),
                })

        processed_event = {
            "event_id": event_id,
            "match_name": f"{home_team} vs {away_team}",
            "league_name": league_name,
            "start_time": start_time.isoformat() if start_time else "Неизвестно",
            "home_team": home_team,
            "away_team": away_team,
            "outcomes": outcomes,
        }
        print("[DEBUG] Обработанные данные события:", json.dumps(processed_event, indent=2, ensure_ascii=False))
        return processed_event
    except Exception as e:
        print(f"[ERROR] Ошибка в process_match_data: {e}")
        print(f"Данные события: {json.dumps(event_data, indent=2, ensure_ascii=False)}")
        return None


# Функция для получения коэффициентов
async def get_live_odds(api: PinnacleAPI):
    url = api._url("v1/odds", {
        "sportId": SPORT_ID,
        "isLive": LIVE_MODE,
        "oddsFormat": ODDS_FORMAT
    })
    data = await api._query(url)

    if not data:
        print("[ERROR] Ответ от API пустой или содержит ошибку.")
        return

    print("[LOG] Ответ от API успешно получен.")
    if "leagues" not in data:
        print("[ERROR] Ключ 'leagues' отсутствует в ответе.")
        print("Ответ:", json.dumps(data, indent=2, ensure_ascii=False))
        return

    for league in data["leagues"]:
        league_name = league.get("name", "Неизвестная лига")
        print(f"Лига: {league_name}")

        if "events" not in league:
            print(f"[WARNING] В лиге '{league_name}' отсутствуют события.")
            continue

        for event in league.get("events", []):
            processed_event = process_match_data(event)
            if not processed_event:
                print("[WARNING] Не удалось обработать событие.")
                continue

            print(f"Матч: {processed_event['match_name']}")
            print(f"Время начала: {processed_event['start_time']}")
            for outcome in processed_event["outcomes"]:
                print(outcome)
            print("-" * 50)


# Основной цикл
async def main():
    api = PinnacleAPI(PINNACLE_USERNAME, PINNACLE_PASSWORD, PROXY)
    while True:
        try:
            await fetch_matches(api)
            await get_live_odds(api)
        except Exception as e:
            print(f"[ERROR] Ошибка в основном цикле: {e}")
        await asyncio.sleep(5)


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        print("[LOG] Работа программы завершена пользователем.")
    except Exception as e:
        print(f"[ERROR] Ошибка в программе: {e}")

File: vovapin.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"net/url"
	"strings"
	"sync"
	"time"
)

// Configuration constants
const (
	PINNACLE_API_URL  = "https://api.pinnacle.com/"
	PINNACLE_USERNAME = "AG1677099"
	PINNACLE_PASSWORD = "5421123A"
	SPORT_ID          = 29
	LIVE_MODE         = 1
	ODDS_FORMAT       = "Decimal"
	REQUEST_TIMEOUT   = 5 * time.Second
	PROXY             = "http://AllivanService:PinnacleProxy@154.7.188.227:5242"
)

// Global data storage for matches
var (
	matchesData      = make(map[int]MatchData)
	matchesDataMutex sync.RWMutex
)

// MatchData stores match information
type MatchData struct {
	Home      string
	Away      string
	League    string
	StartTime string
}

// PinnacleAPI for API interaction
type PinnacleAPI struct {
	Username string
	Password string
	Client   *http.Client
}

// NewPinnacleAPI initializes a new PinnacleAPI instance
func NewPinnacleAPI(username, password, proxy string) *PinnacleAPI {
	transport := &http.Transport{}

	if proxy != "" {
		proxyURL, err := url.Parse(proxy)
		if err != nil {
			fmt.Printf("[ERROR] Invalid proxy URL: %v\n", err)
		} else {
			transport.Proxy = http.ProxyURL(proxyURL)
		}
	}

	client := &http.Client{
		Transport: transport,
		Timeout:   REQUEST_TIMEOUT,
	}

	return &PinnacleAPI{
		Username: username,
		Password: password,
		Client:   client,
	}
}

// buildURL constructs the API endpoint URL with parameters
func (api *PinnacleAPI) buildURL(endpoint string, params map[string]string) string {
	u, err := url.Parse(PINNACLE_API_URL + endpoint)
	if err != nil {
		fmt.Printf("[ERROR] Error parsing URL: %v\n", err)
		return ""
	}

	query := u.Query()
	for k, v := range params {
		if v != "" {
			query.Set(k, v)
		}
	}
	u.RawQuery = query.Encode()
	return u.String()
}

// query makes an HTTP GET request to the specified URL
func (api *PinnacleAPI) query(urlStr string) (map[string]interface{}, error) {
	req, err := http.NewRequest("GET", urlStr, nil)
	if err != nil {
		return nil, err
	}

	req.SetBasicAuth(api.Username, api.Password)
	req.Header.Set("Accept", "application/json")
	req.Header.Set("Content-Type", "application/json")

	resp, err := api.Client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	fmt.Printf("[LOG] Request to URL: %s\n", urlStr)
	fmt.Printf("[LOG] Response status: %d\n", resp.StatusCode)

	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var result map[string]interface{}
	if err := json.Unmarshal(body, &result); err != nil {
		return nil, err
	}

	return result, nil
}

// fetchMatches updates the global matchesData with live match information
func fetchMatches(api *PinnacleAPI) error {
	params := map[string]string{
		"sportId": fmt.Sprintf("%d", SPORT_ID),
		"isLive":  fmt.Sprintf("%d", LIVE_MODE),
	}

	urlStr := api.buildURL("v1/fixtures", params)
	data, err := api.query(urlStr)
	if err != nil {
		fmt.Printf("[ERROR] Error fetching matches: %v\n", err)
		return err
	}

	leagues, ok := data["league"].([]interface{})
	if !ok {
		fmt.Println("[ERROR] Failed to load match data.")
		return fmt.Errorf("missing 'league' key in response")
	}

	matchesDataMutex.Lock()
	defer matchesDataMutex.Unlock()
	matchesData = make(map[int]MatchData)

	for _, leagueInterface := range leagues {
		leagueMap, ok := leagueInterface.(map[string]interface{})
		if !ok {
			continue
		}
		leagueName, _ := leagueMap["name"].(string)
		if leagueName == "" {
			leagueName = "Неизвестная лига"
		}

		events, ok := leagueMap["events"].([]interface{})
		if !ok {
			continue
		}

		for _, eventInterface := range events {
			eventMap, ok := eventInterface.(map[string]interface{})
			if !ok {
				continue
			}

			liveStatus, _ := eventMap["liveStatus"].(float64)
			if int(liveStatus) == 1 {
				eventIDFloat, _ := eventMap["id"].(float64)
				eventID := int(eventIDFloat)
				homeTeam, _ := eventMap["home"].(string)
				if homeTeam == "" {
					homeTeam = "Неизвестная команда"
				}
				awayTeam, _ := eventMap["away"].(string)
				if awayTeam == "" {
					awayTeam = "Неизвестная команда"
				}
				startTime, _ := eventMap["starts"].(string)

				matchesData[eventID] = MatchData{
					Home:      homeTeam,
					Away:      awayTeam,
					League:    leagueName,
					StartTime: startTime,
				}
			}
		}
	}

	return nil
}

// Outcome represents a betting outcome
type Outcome interface {
	String() string
}

// MoneylineOutcome represents a moneyline outcome
type MoneylineOutcome struct {
	Market string
	Team   string
	Odds   float64
}

func (o MoneylineOutcome) String() string {
	return fmt.Sprintf("%s - %s: %.2f", o.Market, o.Team, o.Odds)
}

// TotalOutcome represents a total points outcome
type TotalOutcome struct {
	Market    string
	Line      float64
	OverOdds  float64
	UnderOdds float64
}

func (o TotalOutcome) String() string {
	return fmt.Sprintf("%s - Over %.2f: %.2f, Under %.2f: %.2f", o.Market, o.Line, o.OverOdds, o.Line, o.UnderOdds)
}

// SpreadOutcome represents a spread outcome
type SpreadOutcome struct {
	Market    string
	Line      float64
	HomeOdds  float64
	AwayOdds  float64
}

func (o SpreadOutcome) String() string {
	return fmt.Sprintf("%s - Home %.2f: %.2f, Away %.2f: %.2f", o.Market, o.Line, o.HomeOdds, -o.Line, o.AwayOdds)
}

// TeamTotalOutcome represents a team total outcome
type TeamTotalOutcome struct {
	Market    string
	Team      string
	Line      float64
	OverOdds  float64
	UnderOdds float64
}

func (o TeamTotalOutcome) String() string {
	return fmt.Sprintf("%s (%s) - Over %.2f: %.2f, Under %.2f: %.2f", o.Market, o.Team, o.Line, o.OverOdds, o.Line, o.UnderOdds)
}

// ProcessedEvent stores processed event data
type ProcessedEvent struct {
	EventID    int
	MatchName  string
	LeagueName string
	StartTime  string
	HomeTeam   string
	AwayTeam   string
	Outcomes   []Outcome
}

// processMatchData processes event data and extracts outcomes
func processMatchData(eventData map[string]interface{}) (*ProcessedEvent, error) {
	defer func() {
		if r := recover(); r != nil {
			fmt.Printf("[ERROR] Panic in processMatchData: %v\n", r)
			fmt.Printf("Event data: %v\n", eventData)
		}
	}()

	eventIDFloat, ok := eventData["id"].(float64)
	if !ok {
		return nil, fmt.Errorf("event ID not found")
	}
	eventID := int(eventIDFloat)

	matchesDataMutex.RLock()
	matchData, exists := matchesData[eventID]
	matchesDataMutex.RUnlock()

	var homeTeam, awayTeam, leagueName, startTime string

	if exists {
		homeTeam = matchData.Home
		awayTeam = matchData.Away
		leagueName = matchData.League
		startTime = matchData.StartTime
	} else {
		fmt.Printf("[WARNING] Event %d not found in matchesData.\n", eventID)
		homeTeam = "Неизвестная команда"
		awayTeam = "Неизвестная команда"
		leagueName = "Неизвестная лига"
		startTime = ""
	}

	outcomes := []Outcome{}

	periods, _ := eventData["periods"].([]interface{})
	for _, periodInterface := range periods {
		periodMap, ok := periodInterface.(map[string]interface{})
		if !ok {
			continue
		}

		statusFloat, _ := periodMap["status"].(float64)
		if int(statusFloat) != 1 {
			continue
		}

		periodNumberFloat, _ := periodMap["number"].(float64)
		periodNumber := int(periodNumberFloat)

		periodName := "Матч"
		if periodNumber == 1 {
			periodName = "Первый тайм"
		} else if periodNumber == 2 {
			periodName = "Второй тайм"
		}

		// Moneyline
		if moneyline, ok := periodMap["moneyline"].(map[string]interface{}); ok {
			if odds, exists := moneyline["home"]; exists {
				homeOdds, _ := odds.(float64)
				outcomes = append(outcomes, MoneylineOutcome{
					Market: fmt.Sprintf("Победа (%s)", periodName),
					Team:   "Home",
					Odds:   homeOdds,
				})
			}
			if odds, exists := moneyline["draw"]; exists {
				drawOdds, _ := odds.(float64)
				outcomes = append(outcomes, MoneylineOutcome{
					Market: fmt.Sprintf("Победа (%s)", periodName),
					Team:   "Draw",
					Odds:   drawOdds,
				})
			}
			if odds, exists := moneyline["away"]; exists {
				awayOdds, _ := odds.(float64)
				outcomes = append(outcomes, MoneylineOutcome{
					Market: fmt.Sprintf("Победа (%s)", periodName),
					Team:   "Away",
					Odds:   awayOdds,
				})
			}
		}

		// Totals
		if totals, ok := periodMap["totals"].([]interface{}); ok {
			for _, totalInterface := range totals {
				totalMap, ok := totalInterface.(map[string]interface{})
				if !ok {
					continue
				}

				points, _ := totalMap["points"].(float64)
				overOdds, _ := totalMap["over"].(float64)
				underOdds, _ := totalMap["under"].(float64)

				outcomes = append(outcomes, TotalOutcome{
					Market:    fmt.Sprintf("Общий тотал (%s)", periodName),
					Line:      points,
					OverOdds:  overOdds,
					UnderOdds: underOdds,
				})
			}
		}

		// Spreads
		if spreads, ok := periodMap["spreads"].([]interface{}); ok {
			for _, spreadInterface := range spreads {
				spreadMap, ok := spreadInterface.(map[string]interface{})
				if !ok {
					continue
				}

				hdp, _ := spreadMap["hdp"].(float64)
				homeOdds, _ := spreadMap["home"].(float64)
				awayOdds, _ := spreadMap["away"].(float64)

				outcomes = append(outcomes, SpreadOutcome{
					Market:   fmt.Sprintf("Гандикап (%s)", periodName),
					Line:     hdp,
					HomeOdds: homeOdds,
					AwayOdds: awayOdds,
				})
			}
		}

		// Team totals
		if teamTotal, ok := periodMap["teamTotal"].(map[string]interface{}); ok {
			for teamKey, totalInterface := range teamTotal {
				totalMap, ok := totalInterface.(map[string]interface{})
				if !ok {
					continue
				}

				points, _ := totalMap["points"].(float64)
				overOdds, _ := totalMap["over"].(float64)
				underOdds, _ := totalMap["under"].(float64)

				teamName := "Home"
				if teamKey == "away" {
					teamName = "Away"
				}

				outcomes = append(outcomes, TeamTotalOutcome{
					Market:    fmt.Sprintf("Индивидуальный тотал (%s)", periodName),
					Team:      teamName,
					Line:      points,
					OverOdds:  overOdds,
					UnderOdds: underOdds,
				})
			}
		}
	}

	processedEvent := &ProcessedEvent{
		EventID:    eventID,
		MatchName:  fmt.Sprintf("%s vs %s", homeTeam, awayTeam),
		LeagueName: leagueName,
		StartTime:  startTime,
		HomeTeam:   homeTeam,
		AwayTeam:   awayTeam,
		Outcomes:   outcomes,
	}

	return processedEvent, nil
}

// getLiveOdds fetches live odds and processes each event
func getLiveOdds(api *PinnacleAPI) error {
	params := map[string]string{
		"sportId":    fmt.Sprintf("%d", SPORT_ID),
		"isLive":     fmt.Sprintf("%d", LIVE_MODE),
		"oddsFormat": ODDS_FORMAT,
	}

	urlStr := api.buildURL("v1/odds", params)
	data, err := api.query(urlStr)
	if err != nil {
		fmt.Printf("[ERROR] API response is empty or contains an error: %v\n", err)
		return err
	}

	fmt.Println("[LOG] API response successfully received.")

	leagues, ok := data["leagues"].([]interface{})
	if !ok {
		fmt.Println("[ERROR] Key 'leagues' is missing in response.")
		fmt.Printf("Response: %v\n", data)
		return fmt.Errorf("missing 'leagues' key in response")
	}

	for _, leagueInterface := range leagues {
		leagueMap, ok := leagueInterface.(map[string]interface{})
		if !ok {
			continue
		}

		leagueName, _ := leagueMap["name"].(string)
		if leagueName == "" {
			leagueName = "Неизвестная лига"
		}
		fmt.Printf("Лига: %s\n", leagueName)

		events, ok := leagueMap["events"].([]interface{})
		if !ok {
			fmt.Printf("[WARNING] В лиге '%s' отсутствуют события.\n", leagueName)
			continue
		}

		for _, eventInterface := range events {
			eventMap, ok := eventInterface.(map[string]interface{})
			if !ok {
				continue
			}

			processedEvent, err := processMatchData(eventMap)
			if err != nil {
				fmt.Printf("[WARNING] Не удалось обработать событие: %v\n", err)
				continue
			}

			fmt.Printf("Матч: %s\n", processedEvent.MatchName)
			fmt.Printf("Время начала: %s\n", processedEvent.StartTime)
			for _, outcome := range processedEvent.Outcomes {
				fmt.Println(outcome.String())
			}
			fmt.Println(strings.Repeat("-", 50))
		}
	}

	return nil
}

// main function runs the program loop
func main() {
	api := NewPinnacleAPI(PINNACLE_USERNAME, PINNACLE_PASSWORD, PROXY)

	for {
		if err := fetchMatches(api); err != nil {
			fmt.Printf("[ERROR] Error in main loop (fetchMatches): %v\n", err)
		}

		if err := getLiveOdds(api); err != nil {
			fmt.Printf("[ERROR] Error in main loop (getLiveOdds): %v\n", err)
		}

		time.Sleep(5 * time.Second)
	}
}
